### ISSUE TITLE
User data storage method?

### ISSUE THREAD

**Issue by HittmanA** [Newawe/newawe-node#107]
Hey guys! What are we currently using for storage of user data. Passwords, usernames, profile, etc.

**Comment by as-com** [comment]
The Feathers branch is currently set up to use MySQL, although I am considering RethinkDB.

**Comment by HittmanA** [comment]
I would prefer to use MySQL as I already know a good bit of it.

**Comment by bjskistad** [comment]
Same.

**Comment by HittmanA** [comment]
I have an idea on using LocalStorage/Cookies to store the users info and whether or not they are logged in.

**Comment by bjskistad** [comment]
We can't store info in cookies, they're too easy to hack. (We can store session ids, etc. But no passwords, emails, etc.)

**Comment by HittmanA** [comment]
Ok then made we could try my LocalStorage idea?

**Comment by herohamp** [comment]
how does that work? what if one of the users computers is off?

**Comment by bjskistad** [comment]
No idea what that is. XD

**Comment by herohamp** [comment]
@HittmanA how does LocalStorage work? what if one of the users computers is off?

**Comment by bjskistad** [comment]
@herohamp We would only use it for sessions.

**Comment by HittmanA** [comment]
@herohamp What do you mean if one of their computers is off?

**Comment by herohamp** [comment]
What type of data will be stored in LocalStorage?

**Comment by HittmanA** [comment]
Password Username and whether or not they are logged in.

**Comment by herohamp** [comment]
oh ok

**Comment by as-com** [comment]
Feathers can manage the session tokens for you, and they are in the local storage!

**Comment by HittmanA** [comment]
I do not mean session tokens

**Comment by HittmanA** [comment]
Should I try doing it?

**Comment by as-com** [comment]
It is a terrible idea to store a password in the browser's local storage,
because they are unencrypted and insecure. Rather, it is better to have the
server match up session tokens with users.

**Comment by HittmanA** [comment]
The data is stored on the users computer and no other site can access the data. So why does it matter?

**Comment by bjskistad** [comment]
Someone could use remote access on your computer.

**Comment by herohamp** [comment]
that is your fault though

**Comment by bjskistad** [comment]
Yeah. lol :P

**Comment by HittmanA** [comment]
I agree with @herohamp

**Comment by HittmanA** [comment]
And local storage can store up to 5MB

**Comment by as-com** [comment]
You are literally storing passwords in plain text, which can be easily read by other people. People can get the password from the local storage very easily, and get access to all other accounts with the same password that the user has. Remember, the user never sees it as their own fault, rather, they see it as your fault.

**Comment by HittmanA** [comment]
How can the get stuff from your local storage?

**Comment by HittmanA** [comment]
*Without remote access

**Comment by as-com** [comment]
In Chrome, open Developer Tools (F12), click Resources, and open the local storage pane.

**Comment by HittmanA** [comment]
Ok so that means the **USER** can access it.

**Comment by as-com** [comment]
And anyone with physical access. You are betraying the user's trust by hashing the password on the server, but still storing it in plain text on the user's device.

**Comment by HittmanA** [comment]
Have you not seen Safari autofill? Its the same thing. Anyone with physical access could have all the passwords.

**Comment by as-com** [comment]
If another person gets access to the session token, the worse that could happen is the user's Newawe account is compromised (unless the token had expired), but not all other accounts.

**Comment by HittmanA** [comment]
What other accounts?

**Comment by as-com** [comment]
The OS X keychain is encrypted with the user's password, which is where the autofill is stored.

**Comment by HittmanA** [comment]
Yeah but anyone with physical access can still use it.

**Comment by as-com** [comment]
Relevant XKCD: 
![](https://imgs.xkcd.com/comics/password_reuse.png)

TL;DR: Never store passwords in plaintext, EVER.

**Comment by HittmanA** [comment]
True (XD XKCD). Ok lets use MySQL then.

**Comment by as-com** [comment]
Use cookies or local storage to store session tokens on the client side. Use a database on the server side to store actual user data.

**Comment by HittmanA** [comment]
Yeah.

**Comment by HittmanA** [comment]
MySQL storage is already setup for newawe right?

**Comment by as-com** [comment]
The feathers branch has MySQL set up, except Feathers/Sequelize can't do automatic live table structure migration, which has led me to consider NoSQL solutions like RethinkDB. Think of NoSQL just like SQL, but with JSON-like storage features.

**Comment by HittmanA** [comment]
I hate JSON. To much mental parsing I have to do.

**Comment by as-com** [comment]
Don't be scared of JSON! It's the same JavaScript you know and love...

**Comment by HittmanA** [comment]
Lol! Guess ill give it a second chance.

**Comment by masterex1000** [comment]
what is wrong with the current data method?

**Comment by Radical-from-Scratch** [comment]
We are using HostGator, which has MySQL

**Comment by as-com** [comment]
One does not simply run Node.js on Hostgator…

**Comment by herohamp** [comment]
If you are using HostGator just for the SQL databases I can host you one!

**Comment by herohamp** [comment]
http://www.hostingadvice.com/blog/where-to-find-free-node-js-hosting/

or http://www.evennode.com/pricing ?

**Comment by as-com** [comment]
Node.js hosting is a very messy topic - often, it's better just to manage
your own server or VPS on something like DigitalOcean.

**Comment by herohamp** [comment]
ok yeah lets use digital ocean

**Comment by herohamp** [comment]
what plan?

**Comment by herohamp** [comment]
https://www.digitalocean.com/pricing/

**Comment by HittmanA** [comment]
Just use the $5 plan.

**Comment by herohamp** [comment]
ok

**Comment by Radical-from-Scratch** [comment]
@as-com Yes they do. I am willing to hear your opinions about DigitalOcean, but --- hosting our own? Really? That doesn't seem right.

@HittmanA @heroahamp Baran is paying for it

**Comment by HittmanA** [comment]
Really he is?

**Comment by HittmanA** [comment]
Can we not donate or help out?

**Comment by herohamp** [comment]
I can help pay for i is what I am saying

**Comment by as-com** [comment]
Apps like Newawe can be quite expensive to host. You need to have enough
disk space for a potentially massive amount of user data.

While a $5 digital ocean VPS will suffice for some time, you may run into
memory, CPU, and/or bandwidth issues.

**Comment by Radical-from-Scratch** [comment]
@HittmanA Nobody said that! You are totally welcome to help fund.

**Comment by bjskistad** [comment]
@HittmanA Yes, you will be able to donate. :) Right now, you would just have to send a check, but we can set up a PayPal account for Newawe.

**Comment by bjskistad** [comment]
@as-com Speaking of apps, are we going to have a mobile app?

**Comment by Radical-from-Scratch** [comment]
Um, we *could*, but it would not be until waaaaaaaaaaaaaaaaaay after the main release (we agreed on releasing it on May 14th!!)

**Comment by bjskistad** [comment]
Yeah.

**Comment by bjskistad** [comment]
We need to start working on profiles.

**Comment by Radical-from-Scratch** [comment]
Actually, I live in a state with no sales tax, so I should run da paypal account.

**Comment by bjskistad** [comment]
I'm owner. I own it. Sales tax, so what? I'm a kid and make 40$ a month.

**Comment by bjskistad** [comment]
I'm owner. I own it. Sales tax, so what? I'm a kid and make 40$ a month.

**Comment by herohamp** [comment]
xD

**Comment by Radical-from-Scratch** [comment]
Eh, alright. It was just an offer.

**Comment by bjskistad** [comment]
Does PayPal need money? I believe you need to spend to use it..

**Comment by Radical-from-Scratch** [comment]
I think so. Buy a piece of candy for like 50 cents so you've technically spent money :P

**Comment by bjskistad** [comment]
Lol

**Comment by herohamp** [comment]
ok good

**Comment by herohamp** [comment]
Yeah I can donate like 5 dollars next week

**Comment by as-com** [comment]
Don't worry about money right now. Finish the actual core functionality
first in the Feathers branch. I'm going to set up some more stuff when I
get home.

**Comment by bjskistad** [comment]
K

**Comment by Radical-from-Scratch** [comment]
Also, I found out I am not going to scratch Day so we have allnthe time we need.

**Comment by masterex1000** [comment]
if the relese is on the 14 then a lot of work needs to be done!

**Comment by Radical-from-Scratch** [comment]
Also, I am going to Scratch Day, but we are so not going to release this in 5 days.

**Comment by bjskistad** [comment]
Lol

**Comment by masterex1000** [comment]
Okay.
